std/experimental/checkedint.d

```diff
@@ -0,0 +1,1626 @@
+    if (isIntegral!Rhs || isFloatingPoint!Rhs || is(Rhs == bool))
+    {
+        static assert(is(typeof(mixin("payload"~op~"=rhs")) == T));
+
+        static if (hasMember!(Hook, "hookOpOpAssign"))
+        {
+            hook.hookOpOpAssign!op(payload, rhs);
+        }
+        else
+        {
+            alias R = typeof(payload + rhs);
+            auto r = mixin("this"~op~"rhs").payload;
+
+            static if (valueConvertible!(R, T) ||
+                !hasMember!(Hook, "onBadOpOpAssign") ||
+                op.among(">>", ">>>"))
+            {
+                // No need to check these
+                payload = cast(T) r;
+            }
+            else
+            {
+                static if (isUnsigned!T && !isUnsigned!R)
+                    // Example: ushort += int
+                    const bad = r < 0 || r > max.payload;
+                else
+                    // Some narrowing is afoot
+                    static if (R.min < min.payload)
+                        // Example: int += long
+                        const bad = r > max.payload || r < min.payload;
+                    else
+                        // Example: uint += ulong
+                        const bad = r > max.payload;
+                if (bad)
+                    payload = hook.onBadOpOpAssign!op(payload, Rhs(rhs));
+                else
+                    payload = cast(T) r;
+            }
+        }
+        return this;
+    }
+}
+
+// representation
+unittest
+{
+    assert(Checked!(ubyte, void)(ubyte(22)).representation == 22);
+}
+
+/**
+Force all overflows to fail with `assert(0)`.
+*/
+struct Croak
+{
+static:
+    Dst onBadCast(Dst, Src)(Src src)
+    {
+        assert(0, "Bad cast");
+    }
+    Lhs onBadOpOpAssign(string x, Lhs, Rhs)(Lhs, Rhs)
+    {
+        assert(0, "Bad opAssign");
+    }
+    bool onBadOpEquals(Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        assert(0, "Bad comparison for equality");
+    }
+    bool onBadOpCmp(Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        assert(0, "Bad comparison for ordering");
+    }
+    typeof(~Lhs()) onOverflow(string op, Lhs)(Lhs lhs)
+    {
+        assert(0);
+    }
+    typeof(Lhs() + Rhs()) onOverflow(string op, Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        assert(0);
+    }
+}
+
+unittest
+{
+    Checked!(int, Croak) x;
+    x = 42;
+    short x1 = cast(short) x;
+    //x += long(int.max);
+}
+
+// ProperCompare
+/**
+
+Implements a hook that provides arithmetically correct comparisons for equality
+and ordering. Comparing an object of type $(D Checked!(X, ProperCompare))
+against another integral (for equality or ordering) ensures that no surprising
+conversions from signed to unsigned integral occur before the comparison. Using
+$(D Checked!(X, ProperCompare)) on either side of a comparison for equality
+against a floating-point number makes sure the integral can be properly
+converted to the floating point type, thus making sure equality is transitive.
+
+*/
+struct ProperCompare
```

Comment:
> lmk if better combination is needed.

I don't see any examples of how to compose it with, for example, `Croak`. I think the utility and expected behaviour of that combination are fairly obvious. I know that I could write a whole new `Hook` for that purpose, but surely there ought to be an easier way?

For example, parameterize `ProperCompare` so that I can just write `ProperCompare!Croak`.